src/tests/cwrap/common_mock_nss_dl_load.c

```diff
@@ -0,0 +1,118 @@
+/*
+    Authors:
+        Iker Pedrosa <ipedrosa@redhat.com>
+
+    Copyright (C) 2021 Red Hat
+
+    SSSD tests: Fake nss dl load
+
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation; either version 3 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <dlfcn.h>
+
+#include "common_mock_nss_dl_load.h"
+
+#define LIB_PATH    "libnss_wrapper.so.0"
+
+
+static errno_t load_function(const char *fname, void **fptr)
+{
+    void *dl_handle = NULL;
+    errno_t ret;
+
+    dl_handle = dlopen(LIB_PATH, RTLD_NOW);
```

Comment:
What happens when the same shared object is loaded multiple times?

Wouldn't it be safer to make `dl_handle` static and load library only once?

Reply:
As long as `dlclose()` isn't called, then the same object handle is returned. So, the library is loaded only once.

Reference `man dlopen`:
```
If the same shared object is opened again with dlopen(), the same object handle is returned.
The dynamic linker maintains reference counts for object handles, so a dynamically loaded
shared object is not deallocated until dlclose() has been called on it as many times as dlopen()
has succeeded on it.
```